File: Learn_Basic/src/learn_py/learn_py/topic_opencv/opencv_find_point_publisher.py
import cv2
import numpy as np
import rclpy
from rclpy.node import Node
from learn_interface.msg import TopicOpencv
from geometry_msgs.msg import Point
from cv_bridge import CvBridge

def Find_Point(origin):
    # 将图像从BGR转换为HSV颜色空间
    hsv = cv2.cvtColor(origin, cv2.COLOR_BGR2HSV)

    # 定义红色的HSV范围
    hsv_lower = np.array([156, 43, 46])
    hsv_upper = np.array([180, 255, 255])
    mask = cv2.inRange(hsv, hsv_lower, hsv_upper)
    kernel = np.ones((2, 2), np.uint8)
    mask_opened = cv2.morphologyEx(mask, cv2.MORPH_OPEN, kernel)
    contours, _ = cv2.findContours(mask_opened, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_NONE)

    # 处理查找结果
    radius_max = 0
    point_result = []
    for cnt in contours:
        epsilon = 0.1 * cv2.arcLength(cnt, True)
        approx = cv2.approxPolyDP(cnt, epsilon, True)
        # 拟合
        (x, y), radius = cv2.minEnclosingCircle(approx)
        if radius_max < radius:
            radius_max = radius
            point_result.append((int(x), int(y), int(radius)))  # 添加半径到结果中
            
    return radius_max != 0, point_result


def draw_point_circle(image, posArray):
    for pos in posArray:
        center = (pos[0], pos[1])
        radius = pos[2]  # 使用传入的半径
        cv2.circle(image, center, radius, (255, 0, 0), 2)
        
class PublisherNode(Node):
    def __init__(self, name):
        super().__init__(name)
        self.publisher_ = self.create_publisher(TopicOpencv, "learn_topic_opencv_find_point", 10)
        self.bridge = CvBridge()
        
    def pub(self, data):
        msg = TopicOpencv()
        msg.image = self.bridge.cv2_to_imgmsg(data[0], encoding="bgr8")
        
        # 转换点数据
        point_msgs = []
        for point in data[1]:
            p_msg = Point()
            p_msg.x = float(point[0])
            p_msg.y = float(point[1])
            p_msg.z = float(point[2])
            point_msgs.append(p_msg)
        
        msg.points = point_msgs
        self.publisher_.publish(msg)

def main(args = None):
    
    rclpy.init(args=args)
    
    node = PublisherNode("opencv_find_point_publisher_node")
    
    cap = cv2.VideoCapture(0)

    while True:
        ret, frame = cap.read()
        if not ret:
            break
        try:
            isFind, points = Find_Point(frame)
            
            pub_points = []
            
            for point in points:
                if(point[2] > 10):
                    pub_points.append(point)
            
            sendmessage = [frame, pub_points]
            
            if isFind:
                draw_point_circle(frame, pub_points)
                
            
            node.pub(sendmessage)

        except Exception as e:
            print(e)  # 打印异常信息
            continue

    # 释放摄像头资源
    cap.release()

    rclpy.shutdown()

File: Learn_Basic/src/learn_cmake/src/topic_opencv/opencv_find_point_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "learn_interface/msg/topic_opencv.hpp"
#include "geometry_msgs/msg/point.hpp"
#include "sensor_msgs/msg/image.hpp"
#include "cv_bridge/cv_bridge.hpp"

#include <opencv2/opencv.hpp>
#include <vector>
#include <tuple>
#include <memory>

using namespace std::chrono_literals;

std::pair<bool, std::vector<std::tuple<int, int, int>>> Find_Point(const cv::Mat &origin)
{
    cv::Mat hsv;
    cv::cvtColor(origin, hsv, cv::COLOR_BGR2HSV);

    cv::Scalar hsv_lower(156, 43, 46);
    cv::Scalar hsv_upper(180, 255, 255);
    cv::Mat mask;
    cv::inRange(hsv, hsv_lower, hsv_upper, mask);

    cv::Mat kernel = cv::Mat::ones(2, 2, CV_8U);
    cv::Mat mask_opened;
    cv::morphologyEx(mask, mask_opened, cv::MORPH_OPEN, kernel);

    std::vector<std::vector<cv::Point>> contours;
    cv::findContours(mask_opened, contours, cv::RETR_EXTERNAL, cv::CHAIN_APPROX_NONE);

    int radius_max = 0;
    std::vector<std::tuple<int, int, int>> point_result;

    for (auto &cnt : contours)
    {
        double epsilon = 0.1 * cv::arcLength(cnt, true);
        std::vector<cv::Point> approx;
        cv::approxPolyDP(cnt, approx, epsilon, true);

        cv::Point2f center;
        float radius;
        cv::minEnclosingCircle(approx, center, radius);

        if (radius_max < radius)
        {
            radius_max = static_cast<int>(radius);
            point_result.push_back({static_cast<int>(center.x), static_cast<int>(center.y), static_cast<int>(radius)});
        }
    }

    return {radius_max != 0, point_result};
}

void draw_point_circle(cv::Mat &image, const std::vector<std::tuple<int, int, int>> &posArray)
{
    for (auto &pos : posArray)
    {
        cv::Point center(std::get<0>(pos), std::get<1>(pos));
        int radius = std::get<2>(pos);
        cv::circle(image, center, radius, cv::Scalar(255, 0, 0), 2);
    }
}

class PublisherNode : public rclcpp::Node
{
public:
    PublisherNode(const std::string &name)
        : Node(name)
    {
        publisher_ = this->create_publisher<learn_interface::msg::TopicOpencv>(
            "learn_topic_opencv_find_point", 10);

        cap_.open(0);
        if (!cap_.isOpened())
        {
            RCLCPP_ERROR(this->get_logger(), "Error opening video capture");
        }

        timer_ = this->create_wall_timer(50ms, std::bind(&PublisherNode::timer_callback, this));
    }

private:
    void timer_callback()
    {
        cv::Mat frame;
        cap_ >> frame;
        if (frame.empty())
            return;

        try
        {
            auto [isFind, points] = Find_Point(frame);

            std::vector<std::tuple<int, int, int>> pub_points;
            for (auto &point : points)
            {
                if (std::get<2>(point) > 10)
                    pub_points.push_back(point);
            }

            if (isFind)
            {
                draw_point_circle(frame, pub_points);
            }

            // 构造消息
            learn_interface::msg::TopicOpencv msg;
            auto image_ptr = cv_bridge::CvImage(std_msgs::msg::Header(), "bgr8", frame).toImageMsg();
            msg.image = *image_ptr;

            std::vector<geometry_msgs::msg::Point> point_msgs;
            for (auto &point : pub_points)
            {
                geometry_msgs::msg::Point p;
                p.x = static_cast<double>(std::get<0>(point));
                p.y = static_cast<double>(std::get<1>(point));
                p.z = static_cast<double>(std::get<2>(point));
                point_msgs.push_back(p);
            }
            msg.points = point_msgs;

            publisher_->publish(msg);
        }
        catch (const cv::Exception &e)
        {
            RCLCPP_ERROR(this->get_logger(), "OpenCV exception: %s", e.what());
        }
    }

    rclcpp::Publisher<learn_interface::msg::TopicOpencv>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;
    cv::VideoCapture cap_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<PublisherNode>("opencv_find_point_publisher_node");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
